File: Programmers/level2/거리두기확인하기.cpp
#include<bits/stdc++.h>

using namespace std;
bool checkPartition(pair<int,int> pos1,pair<int,int> pos2,vector<string> board){
    bool result = false;
    int r1 = pos1.first;
    int r2 = pos2.first;
    int c1 = pos1.second;
    int c2 = pos2.second;
    
    if(r1 == r2 && c1 > c2){
        if(board[r1][c1-1] == 'X') result = true;
    }
    else if(r1 == r2 && c1 < c2){
        if(board[r1][c1+1] == 'X') result= true;
    }
    else if(c1 == c2 && r1 > r2){
        if(board[r1-1][c1] == 'X') result = true;
    }
    else if(c1 == c2 && r1 < r2){
        if(board[r1+1][c1] == 'X') result = true;
    }
    
    else if(c1 < c2 && r1 < r2){
        if(board[r1+1][c1] == 'X' && board[r1][c1+1] == 'X') result = true;
    }
    else if(c1 > c2 && r1 > r2){
        if(board[r1-1][c1] == 'X' && board[r1][c1-1] == 'X') result = true;
    }
    else if(c1 < c2 && r1 > r2){
        if(board[r1-1][c1] == 'X' && board[r1][c1+1] == 'X') result = true;
    }
    else if(c1 > c2 && r1 < r2){
        if(board[r1+1][c1] == 'X' && board[r1][c1-1] == 'X') result = true;
    }
    return result;
}

int getDist(pair<int,int> pos1,pair<int,int> pos2){
    int r1 = pos1.first;
    int r2 = pos2.first;
    int c1 = pos1.second;
    int c2 = pos2.second;
    return abs(r2-r1) + abs(c2-c1);
}

bool solution(vector<string> board){
    bool result = true;
    vector<pair<int,int>> candi;
    for(int i=0;i<5;i++){
        for(int j=0;j<5;j++){
            if(board[i][j] == 'P')
                candi.push_back({i,j});
        }
    }
    int sz = candi.size();
    for(int i=0;i<sz-1;i++){
        for(int j=i+1;j<sz;j++){
            int dist = getDist(candi[i],candi[j]);
            if(dist == 1) {
                return false;
            }
            else if(dist == 2){
                bool check = checkPartition(candi[i],candi[j],board);
                if(!check){
                    return false;
                }
            }
            
        }
    }
    return result;
        
}
vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    for(auto board : places){
        answer.push_back(solution(board));
    }
    return answer;
}

File: Java/Programmers/level2/거리두기확인하기.java
import java.util.*;
class Solution {
    class Pair{
        int first;
        int second;
        public Pair(int a,int b){
            first = a;
            second = b;
        }
    }
    public int getDist(int r1,int c1,int r2,int c2){
        return Math.abs(r1-r2) + Math.abs(c1-c2);
    }
    
    public boolean check(int r1,int c1,int r2, int c2,String [] place){
        if(r1 == r2 && c1 > c2 ){ // 왼쪽
            if(place[r1].charAt(c1-1) == 'X') return true;
            else return false;
        } 
        else if(r1 < r2 && c1 > c2){ // 왼쪽 대각
            if(place[r1].charAt(c1-1) == 'X' && place[r1+1].charAt(c1) == 'X') return true;
            else return false;
        }
        else if(r2 > r1 && c1 == c2){ // 밑
            if(place[r1+1].charAt(c1) == 'X') return true;
            else return false;
        }
        else if(r2 > r1 && c2>c1){ //오른쪽 대각
            if(place[r1+1].charAt(c1+0) == 'X' && place[r1].charAt(c1+1) == 'X') return true;
            else return false;
        }
        else if(r1 == r2 && c1 < c2){ // 오른쪽
            if(place[r1].charAt(c1+1) == 'X') return true;
            else return false;
        }
        else{
            System.out.println("incorrect ranget");
            return false;
        } 
    }
    
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        int tc = -1;
        
        while(++tc < 5 ){
            boolean isSafe = true;
            List<Pair> v = new ArrayList<>();
            for(int i=0;i<5;i++){ //사람이 들어있는 좌표 리스트에 저장
                for(int j=0;j<5;j++)
                    if(places[tc][i].charAt(j) == 'P') v.add(new Pair(i,j));
            }
            int sz = v.size();
            for(int i = 0;i<sz-1;i++){
                Pair pos1 = v.get(i);
                for(int j=i+1;j<sz;j++){
                    Pair pos2 = v.get(j);
                    int dist = getDist(pos1.first,pos1.second,pos2.first,pos2.second);
                    if(dist == 1){
                        isSafe = false;
                        break;
                    }
                    else if(dist == 2){
                        if(!check(pos1.first,pos1.second,pos2.first,pos2.second,places[tc])){
                            isSafe = false;
                            break;
                        }
                    }
                }
                if(!isSafe) break;
            }
            
            answer[tc] = isSafe ? 1 : 0;
        }
        
        return answer;
    }
}